src/DomainHost.sol

```diff
@@ -0,0 +1,427 @@
+    function frob(bytes32 i, address u, address v, address w, int dink, int dart) external;
+    function suck(address u, address v, uint256 rad) external;
+    function urns(bytes32, address) external view returns (uint256, uint256);
+    function grab(bytes32, address, address, address, int256, int256) external;
+}
+
+interface DaiJoinLike {
+    function vat() external view returns (VatLike);
+    function dai() external view returns (DaiLike);
+    function join(address usr, uint256 wad) external;
+    function exit(address usr, uint256 wad) external;
+}
+
+interface DaiLike {
+    function transfer(address dst, uint256 wad) external returns (bool);
+    function transferFrom(address src, address dst, uint256 wad) external returns (bool);
+    function approve(address usr, uint wad) external returns (bool);
+}
+
+interface RouterLike {
+    function settle(bytes32 sourceDomain, bytes32 targetDomain, uint256 amount) external;
+    function registerMint(TeleportGUID calldata teleport) external;
+}
+
+struct Settlement {
+    bytes32 sourceDomain;
+    bytes32 targetDomain;
+    uint256 amount;
+    bool    sent;
+}
+
+/// @title Support for xchain MCD, canonical DAI and Maker Teleport - host instance
+/// @dev This is just the business logic which needs concrete message-passing implementation
+abstract contract DomainHost {
+
+    // --- Data ---
+    mapping (address => uint256) public wards;
+    mapping (bytes32 => bool)    public teleports;
+    Settlement[]                 public settlementQueue;
+
+    address public vow;
+    uint256 public lid;         // Local ordering id
+    uint256 public rid;         // Remote ordering id
+    uint256 public line;        // Remote domain global debt ceiling [RAD]
+    uint256 public grain;       // Keep track of the pre-minted DAI in the escrow [WAD]
+    uint256 public sin;         // A running total of how much is required to re-capitalize the remote domain [WAD]
+    uint256 public cure;        // The amount of unused debt [RAD]
+    bool public cureReported;   // Returns true if cure has been reported by the guest
+    uint256 public live;
+
+    bytes32     public immutable ilk;
+    VatLike     public immutable vat;
+    DaiJoinLike public immutable daiJoin;
+    DaiLike     public immutable dai;
+    address     public immutable escrow;
+    RouterLike  public immutable router;
+
+    uint256 constant RAY = 10 ** 27;
+
+    string constant ARITHMETIC_ERROR = string(abi.encodeWithSignature("Panic(uint256)", 0x11));
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed what, address data);
+    event Lift(uint256 wad);
+    event Release(uint256 wad);
+    event Push(int256 wad);
+    event Rectify(uint256 wad);
+    event Cage();
+    event Tell(uint256 value);
+    event Exit(address indexed usr, uint256 wad);
+    event UndoExit(address indexed originalSender, uint256 wad);
```

Comment:
Shouldn't there be a correspondence between `Exit` and `UndoExit` events? I mean it should be possible whether given exit was undone or not by looking for corresponding UndoExit event.